### PR TITLE
Add --force to kubectl delete and explain force deletion

### PR DESCRIPTION
**Pull request by smarterclayton**
--force is required for --grace-period=0. --now is == --grace-period=1.
Improve command help to explain what graceful deletion is and warn about
force deletion.

Part of #34160 & #29033

```release-note
In order to bypass graceful deletion of pods (to immediately remove the pod from the API) the user must now provide the `--force` flag in addition to `--grace-period=0`.  This prevents users from accidentally force deleting pods without being aware of the consequences of force deletion.  Force deleting pods for resources like StatefulSets can result in multiple pods with the same name having running processes in the cluster, which may lead to data corruption or data inconsistency when using shared storage or common API endpoints.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35484)

### CODE REVIEW

**Comment by smarterclayton** [comment]
@kubernetes/kubectl see referenced proposal for description of changes, @foxish

File: pkg/kubectl/cmd/delete.go [hunk 1]
```diff
@@ -148,7 +163,10 @@ func RunDelete(f cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []stri
 		if gracePeriod != -1 {
 			return fmt.Errorf("--now and --grace-period cannot be specified together")
 		}
-		gracePeriod = 0
+		gracePeriod = 1
+	}
+	if gracePeriod == 0 && !cmdutil.GetFlagBool(cmd, "force") {
+		return fmt.Errorf("Immediate deletion may result in multiple copies of your resources running at the same time. You must pass --force to delete with grace period 0.")
```
Comment (foxish) [inline pkg/kubectl/cmd/delete.go]: I think this error message may not be clear to someone who isn't familiar with the entire issue. Also, if it's force deleting a simple pod object, the multiple copies comment does not apply. Maybe we could rephrase it to say something along the lines of:

> Immediate deletion does not wait for confirmation that the running resource has been terminated and may leave it running in case of a network partition.

**Comment by smarterclayton** [comment]
Updated text (but not generated files)

**Comment by smarterclayton** [comment]
@foxish @vishh @bprashanth if you guys have more opinions on the text here please let me know, otherwise going to generate docs and make sure the tests are right.

**Comment by smarterclayton** [comment]
Also any opinionated kubectl reviewers on messages on delete.  This is an explicit CLI API change for safety purposes and can break scripts that were doing unsafe things before.

File: pkg/kubectl/cmd/delete.go [hunk 2]
```diff
@@ -148,7 +165,10 @@ func RunDelete(f cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []stri
 		if gracePeriod != -1 {
 			return fmt.Errorf("--now and --grace-period cannot be specified together")
 		}
-		gracePeriod = 0
+		gracePeriod = 1
+	}
+	if gracePeriod == 0 && !cmdutil.GetFlagBool(cmd, "force") {
+		return fmt.Errorf("Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely. You must pass --force to delete with grace period 0.")
```
Comment (vishh) [inline pkg/kubectl/cmd/delete.go]: nit: I'd place a `WARNING:` prefix to this message!

File: pkg/kubectl/cmd/delete.go [hunk 3]
```diff
@@ -36,9 +36,22 @@ var (
...
-
-		Only one type of the arguments may be specified: filenames, resources and names, or resources and label selector
+		JSON and YAML formats are accepted. Only one type of the arguments may be specified: filenames,
+		resources and names, or resources and label selector.
+
+		Some resources, such as pods, support graceful deletion. These resources define a default period
+		before they are forcibly terminated (the grace period) but you may override that value with
+		the --grace-period flag, or pass --now to set a grace-period of 1. Because these resources often
+		represent entities in the cluster, deletion may not be acknowledged immediately. If a resource
+		like a pod has not been deleted within the grace period, the node that hosts that pod may be
```
Comment (vishh) [inline pkg/kubectl/cmd/delete.go]: Grace period can be as low as 1 second. We should not place expectations on the system to gracefully delete pod objects in 1 second. Instead, let's say that `if a pod isn't deleted in a reasonable amount of time, the node might be down and/or unreachable. In such scenarios, it might be necessary to forcibly delete pods to free up cluster resources.`

File: pkg/kubectl/cmd/delete.go [hunk 4]
```diff
@@ -36,9 +36,22 @@ var (
...
+		down or unable to communicate to the server. To force delete a resource, you must pass a grace
+		period of 0 and specify the --force flag.
+
+		IMPORTANT: Force deleting pods does not wait for confirmation that the pod's processes have been
+		terminated, which means the pod may still be running on the cluster. This may result in
```
Comment (vishh) [inline pkg/kubectl/cmd/delete.go]: Not just concurrent access, it will also lead to resource leakage on the nodes which over time can lead to unexpected evictions.

Comment (smarterclayton) [inline pkg/kubectl/cmd/delete.go]: I don't think most users care about that - that's a kubelet problem.  Most users care about "my database is now corrupted" >>>> "nodes may not have as many resources as they said".

**Comment by foxish** [comment]
@smarterclayton Looks fine to me. 
cc @pwittrock

**Comment by smarterclayton** [comment]
node e2e looks like a flake (or maybe needs a --force and is failing hard somewhere)

```
I1101 20:16:11.753] Connection to 104.154.151.234 closed by remote host.
I1101 20:16:11.753] 
I1101 20:16:11.753] Failure Finished Test Suite on Host tmp-node-e2e-4254901c-gci-dev-55-8872-18-0
I1101 20:16:11.754] [command [ssh -i /home/jenkins/.ssh/google_compute_engine -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o CheckHostIP=no -o StrictHostKeyChecking=no -o ServerAliveInterval=30 -o LogLevel=ERROR 104.154.151.234 -- sh -c 'cd /tmp/gcloud-e2e-404546779 && timeout -k 30s 2700.000000s ./ginkgo --nodes=8 --skip="\[Flaky\]|\[Slow\]|\[Serial\]" --flakeAttempts=2 ./e2e_node.test -- --logtostderr --v 4 --node-name=tmp-node-e2e-4254901c-gci-dev-55-8872-18-0 --report-dir=/tmp/gcloud-e2e-404546779/results --report-prefix=gci-family-gci-dev-55-8872-18-0 --experimental-mounter-path=/tmp/gcloud-e2e-404546779/cluster/gce/gci/mounter/mounter --experimental-mounter-rootfs-path=/media/root '] failed with error: exit status 255 and output:
```

**Comment by soltysh** [comment]
LGTM

**Comment by smarterclayton** [comment]
Any other comments or is this good enough for government work?

File: pkg/kubectl/cmd/delete.go (hunk 2 above)
Comment (vishh) [inline pkg/kubectl/cmd/delete.go]: Excuse my ignorance, why is it not possible to add a string prefix to the error message here?

File: pkg/kubectl/cmd/delete.go (hunk 4 above)
Comment (vishh) [inline pkg/kubectl/cmd/delete.go]: Well, users don't care about resources up and until their pods start getting evicted. Would it hurt conveying this additional information?

**Comment by vishh** [comment]
@smarterclayton two more open comments.

File: pkg/kubectl/cmd/delete.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/kubectl/cmd/delete.go]: I'm just trying not to make this a novel.  Let me see if I can work in a sentence.

**Comment by krousey** [comment]
@smarterclayton 
This is causing upgrade tests to fail. 1.3 e2e tests running with a 1.5 kubectl. I'm open to clever solutions.

cc @saad-ali 

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-e2e-gke-container_vm-1.3-container_vm-1.5-upgrade-cluster/143?log

```console
Expected error:
    <*errors.errorString | 0xc820019820>: {
        s: "Error running &{/workspace/kubernetes_skew/cluster/kubectl.sh [/workspace/kubernetes_skew/cluster/kubectl.sh --server=https://199.223.235.243 --kubeconfig=/workspace/.kube/config delete --grace-period=0 -f - --namespace=e2e-tests-kubectl-up6nc] []  0xc8207b9b60  error: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely. You must pass --force to delete with grace period 0.\n [] <nil> 0xc8206fe440 exit status 1 <nil> true [0xc820ea4718 0xc820ea4740 0xc820ea4750] [0xc820ea4718 0xc820ea4740 0xc820ea4750] [0xc820ea4720 0xc820ea4738 0xc820ea4748] [0xa97840 0xa979a0 0xa979a0] 0xc82027b800}:\nCommand stdout:\n\nstderr:\nerror: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely. You must pass --force to delete with grace period 0.\n\nerror:\nexit status 1\n",
    }
    Error running &{/workspace/kubernetes_skew/cluster/kubectl.sh [/workspace/kubernetes_skew/cluster/kubectl.sh --server=https://199.223.235.243 --kubeconfig=/workspace/.kube/config delete --grace-period=0 -f - --namespace=e2e-tests-kubectl-up6nc] []  0xc8207b9b60  error: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely. You must pass --force to delete with grace period 0.
     [] <nil> 0xc8206fe440 exit status 1 <nil> true [0xc820ea4718 0xc820ea4740 0xc820ea4750] [0xc820ea4718 0xc820ea4740 0xc820ea4750] [0xc820ea4720 0xc820ea4738 0xc820ea4748] [0xa97840 0xa979a0 0xa979a0] 0xc82027b800}:
    Command stdout:
    
    stderr:
    error: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely. You must pass --force to delete with grace period 0.
    
    error:
    exit status 1
    
not to have occurred
```

**Comment by saad-ali** [comment]
> @smarterclayton This is causing upgrade tests to fail. 1.3 e2e tests running with a 1.5 kubectl. I'm open to clever solutions.

Thanks for raising this issue @krousey! Can you create an bug with the 1.5 milestone to track this.  @smarterclayton any thoughts on how to handle this? Can you have server side check client version before enforcing the new behavior?

**Comment by foxish** [comment]
This change is completely at the client-side and is a departure from the old behavior. I imagine one could check the server version and then enforce the use of the `--force` flag, but that sounds icky. 
Is there another option here?

**Comment by krousey** [comment]
@foxish I just created an issue. Please move discussion to there.

**Comment by bgrant0607** [comment]
This breaks all previous uses of `--grace-period=0`? If so, we should revert this PR. 

We need to not break scripts using kubectl where there was no well defined way of ensuring forward compatibility in order to avoid such breakage.

**Comment by smarterclayton** [comment]
The break was intentional and the reasoning laid in the proposal for why allowing this to continue is unsafe.

**Comment by smarterclayton** [comment]
Let's continue discussion in #37117
